llama.cpp

```diff
@@ -948,7 +950,22 @@ struct llama_hparams {
     float rope_freq_scale_train;
 
     bool operator!=(const llama_hparams & other) const {
-        return static_cast<bool>(memcmp(this, &other, sizeof(llama_hparams))); // NOLINT
+        if(this->vocab_only != other.vocab_only) return true;
+        if(this->n_vocab != other.n_vocab) return true;
+        if(this->n_ctx_train != other.n_ctx_train) return true;
+        if(this->n_embd != other.n_embd) return true;
+        if(this->n_head != other.n_head) return true;
+        if(this->n_head_kv != other.n_head_kv) return true;
+        if(this->n_layer != other.n_layer) return true;
+        if(this->n_rot != other.n_rot) return true;
+        if(this->n_ff != other.n_ff) return true;
+        
+        if(std::abs(this->f_norm_eps - other.f_norm_eps) > EPSILON) return true;
+        if(std::abs(this->f_norm_rms_eps - other.f_norm_rms_eps) > EPSILON) return true;
+        if(std::abs(this->rope_freq_base_train - other.rope_freq_base_train) > EPSILON) return true;
+        if(std::abs(this->rope_freq_scale_train - other.rope_freq_scale_train) > EPSILON) return true;
```

Comment:
Is this correct if either value is NaN or +/-inf? We should probably implement our own equivalent to python's math.isclose to make sure we get this right - though using a single epsilon value instead of two should be OK.

Reply:
See https://github.com/python/cpython/blob/77e9aae3837d9f0cf87461d023896f2c4aeb282f/Modules/mathmodule.c#L3109
A few important points:
- It compares for equality to find two infinities of the same sign
- It checks whether either value is infinity
- It checks whether the diff is _less_ than epsilon, which I believe is important in the case of NaN - I think NaN compared to anything is always false.

Reply:
Added a `is_float_eq` helper function that mimics the python `is_close` function but using single epsilon